### PR TITLE
Principal::getForOutput now always retrieves the product

### PR DESCRIPTION
**Pull request by Dr15Jones**
The method Principal::getForOutput now always retrieves the product. If only the provenance is needed one should call ProductProvenanceRetriever directly.
This change is needed as a step towards allowing more than one thread per event.

### CODE REVIEW

**Comment by Dr15Jones** [comment]
This includes the same commit as #13467 since the new commit in this pull request needs that change.

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
@wmtan could you do a review of this change to double check I didn't miss something?

**Comment by Dr15Jones** [comment]
The test failure is unrelated to this pull request and comes from the IB.

**Comment by wmtan** [comment]
@Dr15Jones I just now saw your e-mail asking me to review this. So, I will look at it now.

File: IOPool/Output/src/RootOutputFile.cc [hunk 1]
```diff
@@ -696,6 +694,27 @@ namespace edm {
...
+    ProductProvenanceRetriever const* provRetriever = nullptr;
+    {
+      EventPrincipal const* eventPrincipal = dynamic_cast<EventPrincipal const*>(&principal);
+      if (eventPrincipal) {
+        provRetriever =eventPrincipal->productProvenanceRetrieverPtr();
+        assert(provRetriever);
+      }
+    }
+    
+    //If we are dropping some of the meta data we need to know
```
Comment (wmtan) [inline IOPool/Output/src/RootOutputFile.cc]: @Dr15Jones I don't understand the purpose of lines 707-717, and the use of "producedBranches" after that.

Comment (Dr15Jones) [inline IOPool/Output/src/RootOutputFile.cc]: It is there to replace the call on line 675 that I removed 
`principal.getProvenance(info->branchID(), mcc).product().produced())`

That `getProvenance` call required access to the principal and could trigger reading from the file but that same information can be gotten from the `ProductRegistry`.

**Comment by Dr15Jones** [comment]
+1

**Comment by Dr15Jones** [comment]
@davidlange6 ping?

**Comment by wmtan** [comment]
@davidlange6 Can this PR from Chris be merged? The test failure is unrelated to this pull request and comes from the IB. I'm doing work on top of this that will be much easier after this PR is in an IB.

**Comment by davidlange6** [comment]
yep - sorry, didn't look at PRs this morning.

**Pull request merged**